kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java

```diff
@@ -0,0 +1,494 @@
+
+    public PartitionSegments(
+        final MongoDatabase database,
+        final MongoDatabase adminDatabase,
+        final SessionSegmentPartitioner partitioner,
+        final int kafkaPartition,
+        final long streamTime,
+        final long epoch,
+        final CollectionCreationOptions collectionCreationOptions
+    ) {
+      this.database = database;
+      this.adminDatabase = adminDatabase;
+      this.partitioner = partitioner;
+      this.epoch = epoch;
+      this.collectionCreationOptions = collectionCreationOptions;
+      this.segmentSessions = new ConcurrentHashMap<>();
+
+      final var activeSegments = partitioner.segmenter().activeSegments(kafkaPartition, streamTime);
+      if (activeSegments.isEmpty()) {
+        LOG.info("{}[{}] No active segments for initial streamTime {}",
+                 database.getName(), kafkaPartition, streamTime);
+      } else {
+        for (final Segmenter.SegmentPartition segmentToCreate : activeSegments) {
+          // Most likely if we have active segments upon initialization, the corresponding
+          // collections and windowStart indices already exist, in which case #getCollection and
+          // #createIndex should be quick and won't rebuild either structure if one already exists
+          createSegment(segmentToCreate);
+        }
+
+        final long firstSegmentId = activeSegments.get(0).segmentStartTimestamp;
+        LOG.info("{}[{}] Initialized active segments in range {} - {}",
+                 database.getName(), kafkaPartition, firstSegmentId,
+                 firstSegmentId + activeSegments.size());
+      }
+    }
+
+    private String collectionNameForSegment(final Segmenter.SegmentPartition segmentPartition) {
+      final long segmentStartTimeMs =
+          segmentPartition.segmentStartTimestamp * partitioner.segmenter().segmentIntervalMs();
+      return String.format(
+          "%d-%d",
+          segmentPartition.tablePartition, segmentStartTimeMs
+      );
+    }
+
+    private void createSegment(final Segmenter.SegmentPartition segmentToCreate) {
+      LOG.info("{}[{}] Creating segment id {}",
+          database.getName(), segmentToCreate.tablePartition, segmentToCreate.segmentStartTimestamp
+      );
+
+      final var collectionName = collectionNameForSegment(segmentToCreate);
+      final MongoCollection<SessionDoc> sessionDocs;
+      if (collectionCreationOptions.sharded()) {
+        sessionDocs = MongoUtils.createShardedCollection(
+            collectionName,
+            SessionDoc.class,
+            database,
+            adminDatabase,
+            collectionCreationOptions.numChunks()
+        );
+      } else {
+        sessionDocs = database.getCollection(
+            collectionNameForSegment(segmentToCreate),
+            SessionDoc.class
+        );
+      }
+      segmentSessions.put(segmentToCreate, sessionDocs);
+
+      // TODO(agavra): make the tombstone retention configurable
```

Comment:
I'm still a bit hesitant about whether we want to support it at all for window stores, but as noted already, we clearly need to for KV stores and session stores.

However I don't think there's any reasonable need to configure it per store. Maybe we can just throw in a global config for tombstone retention and call it a day? (Not in this PR of course)  cc @agavra